src/Corax/Queries/SuggestionTermProvider.cs

```diff
@@ -1,140 +1,481 @@
+            {
+                // We advance one place
+                _currentIdx++;
+
+                // Check if we are not done. 
+                if (_currentIdx >= _count)
+                {
+                    boost = 0;
+                    ngram = ReadOnlySpan<byte>.Empty;
+                    return false;
+                }
+
+                ngram = _storage.AsSpan().Slice(_gramPosition[_currentIdx], _gramSize[_currentIdx]);
+                boost = _gramBoost[_currentIdx];
+                return true;
+            }
+        }
+
+
+        public static SuggestionTermProvider<TDistanceProvider> YieldSuggestions(IndexSearcher searcher, int fieldId, Slice term, IndexFieldBinding binding, TDistanceProvider distanceProvider, bool sortByPopularity, float distance, int take = -1)
+        {
+            static void NextNoDistance(ref SuggestionTermProvider<TDistanceProvider> provider, ref Span<byte> terms, ref Span<Token> tokens, ref Span<float> scores)
             {
                 int fieldId = provider._fieldId;
                 var allocator = provider._searcher.Allocator;
-                var term = provider._term;                
 
                 Slice.From(allocator, $"__Suggestion_{fieldId}", out var treeName);
                 var tree = provider._searcher.Transaction.CompactTreeFor(treeName);
 
-                var keys = SuggestionsKeys.Generate(allocator, DefaultNGramSize, term.AsSpan(), out int keysCount).ToReadOnlySpan();
-                int keySizes = keys.Length / keysCount;
+                var gramTable = new SuggestionsNGramTable(DefaultNGramSize);
+                gramTable.Generate(provider._term.AsSpan());
+
+                // We initialize the iterator and store it in the stack memory.
+                var iter = tree.Iterate();
+                var values = new FastList<(Slice Term, int Popularity)>();
+                var dictionary = new Dictionary<uint, int>();
 
-                var values = new FastList<(Slice, int)>();
-                var dictionary = new Dictionary<Slice, int>();                
-                for (int i = 0; i < keysCount; i++)
+                int maxPopularity = 0;
+                while (gramTable.MoveNext(out var ngram, out var boost))
                 {
-                    // We initialize the iterator and store it in the stack memory.
-                    var iter = tree.Iterate();
-                    iter.Seek(keys.Slice(i * keySizes, DefaultNGramSize));
+                    iter.Seek(ngram);
 
-                    byte lastByte = keys[i * keySizes + DefaultNGramSize - 1];
-                    while (iter.MoveNext(out Slice key, out var _))
+                    byte lastByte = ngram[^1];
+                    while (iter.MoveNext(out Slice gramKey, out var _))
                     {
-                        if (key[DefaultNGramSize - 1] != lastByte)
+                        if (gramKey[ngram.Length - 1] != lastByte)
```

Comment:
that is attempting to avoid the comparison costs?
Isn't that dangerous if we have a `gramKey` that is less than `ngram.Length-1` ?

Reply:
That `ngram` key comes from `SuggestionsNGramTable` which ensures that the minimum ngram size is 2 (like in Lucene)